IBM/carbon-addons-iot-react
# 1051 cells can truncate content show tooltip if truncated

Closes #1051

**Summary**

> In data tables, we need to allow the content to truncate so that the rows don't become very very tall if the content is longer that the set column widgth. If the content is trunctated, the user should be able to click on it to see the full, untruncated content.

_Implementation_
![image](https://user-images.githubusercontent.com/5167091/78908451-5f370580-7a82-11ea-9f23-198a9dd893b4.png)

Cell text wrapping is configurable via an option.wrapText prop.

1. If wrapText is 'auto' we apply wrapping only for tables with normal columns*. For tables with fixed or resizable columns we use truncation.
2. if wrapText is 'always' we wrap in all scenarios including tables with fixed or resizable columns (i.e. current situation).
3. if wrapText is 'never' we never wrap and as a consequence the tables with normal columns* get wide cells and more often horizontal scrolling. For tables with fixed or resizable columns we use truncation.

*) table with normal columns is here a table with the default table-layout:auto and with neither column resize nor fixed column widths.

See further discussion [here](https://github.com/IBM/carbon-addons-iot-react/issues/1051)


**Change List (commits, features, bugs, etc)**
- added an external prop called wrapCellText
- added an internal prop called truncateCellText which is calculated in Table.jsx and passed down to the affected table components.
- wrapping like normal (carbon tables) is the default
- Truncation will only happen if wrapText:auto or wrapText:never for `hasResize:true` or if *any* of the columns has a set width.
- Changed min-width of columns so that truncated column header text and sort arrows fit together
- The tooltip is only added for table body cells, not the table header cells. Table header cells have the title showing on hover, but a click to show the carbon tooltip would be strange since click is used for sorting.
- The tooltip is only added if actual truncation is occurring, therefore all cells are measured for that.
- The tooltip will adopt to column width changes made by the user, i.e. if a previously truncated cell will have enough space after the user resizes the column the tooltip will be removed.
- Updated the relevant stories to have a knob for wrapCellText
- Added unit tests for the above described functionality when possible/reasonable

- Fixed a bug where the minimal width of the last column was not respected if the table needed horizontal scrollbars.


**Acceptance Test (how to verify the PR)**
For all cases adjust your browser width to be around 1300px to see the effects of wrapping and truncation.

- [x] Table with column resize

https://deploy-preview-1078--carbon-addons-iot-react.netlify.com//?path=/story/watson-iot-table--with-resize-and-initial-column-widths-on-simple-stateful-and-row-selection

0. Set knob `wrapCellText` to `auto` 
1. Verify that the cell texts are being truncated where needed.
2. Click on the truncated texts to check that the tooltip is showing (also try scrolling when the tooltip is showing)
3. Verify that non truncated texts do not have a tooltip 
4. Resize a column that isn't truncating, e.g Secret Information, and verify that truncation and tooltip is added, also the the reverse.
5. Select `always` for the knob wrapCellTex and make sure the cell texts are now wrapping

- [x] Table with fixed column widths

https://deploy-preview-1078--carbon-addons-iot-react.netlify.com//?path=/story/watson-iot-table--with-fixed-column-width-and-no-resize

0. Set knob `wrapCellText` to `auto` 
1. Verify that the cell texts are being truncated where needed.
2. Select `always` for the knob wrapCellTex and make sure the cell texts are now wrapping

- [x] Table with no column resize nor any fixed width columns (i.e. dynamic column widths)

https://deploy-preview-1078--carbon-addons-iot-react.netlify.com/?path=/story/watson-iot-table--simple-stateful-example

1. Check that column texts are wrapping and that the column widths are distributed proportionally based on the content of the cells.
2. Select `never` for the knob wrapCellText and verify that the table no longer wrap the texts. The table should grow in size and get horizontal scrolling.
